keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,819 @@
+  
+	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
+    // It can return an error if the field selector cannot be parsed.
+    ParseLabelSelector() ([]Requirement, error)
+```
+
+#### Future-proofing your authentication webhook for future verbs
+
+As of 1.31, the only verbs with field and label selectors are List, Watch, and DeleteCollection.
+<<[UNRESOLVED deads2k, liggitt, jpbetz, sttts: this is a future promise for the kube-apiserver behavior if we add it ]>>
+In the future, the kube-apiserver may add field and label selectors to Create, Update, and Delete.
+<<[/UNRESOLVED]>>
+* For Create, this means that the resource after all mutation is complete (finalObject) must match the field and label selector. 
+* For Update, this means that the finalNewObject and oldObject must match the field and label selector.
+* For Delete, this means that the oldObject must match the field and label selector.
+ 
+We do not expand field and label selectors for Get, because if a client is specifying a selector, they can add a `.metadata.name`
+field selector and use a List to get equivalent functionality.
+
+### SubjectAccessReview Changes
+
+SubjectAccessReview is used for two purposes:
+1. Authorization webhook calls from the kube-apiserver to a webhook.
+   This usage likely benefits from a serialization with `[]Requirement`.
+2. Authorization checks from a client (often a server process using in-cluster authorization like kube-rbac-proxy)
+   This usage likely benefits from a serialization that matches the query parameter.
+
+Their needs best met with two different serialization (see user stories)
+
+```go
+type SubjectAccessReviewSpec struct {
+	FieldSelector *FieldSelectorAttributes
+	
+	LabelSelector *LabelSelectorAttributes
+}
+
+// FieldSelectorAttributes indicates a field limited access.
+// For webhooks:
+// The kube-apiserver will never send a request with rawSelector set, but we cannot control what other clients directly send.
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the request is not limited.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the request is invalid.
+// For the kube-apiserver:
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the rawSelector will be parsed and limited if the parsing succeeds.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the request is invalid.
+type FieldSelectorAttributes struct {
+	// rawSelector is the serialization of a field selector that would be included in a List query parameter.
+	// Webhook implementations must handle malformed rawSelectors.
```

Comment:
> 	// Webhook implementations must handle malformed rawSelectors.

Must they? A lot of the docs seem leftover from the first draft where the fields were not mutually exclusive

I'd encourage general webhook authors to:
1. ensure rawSelector and requirements are not both set
2. consider the `requirements` field if set
3. not try to parse or consider the `rawSelector` field if set


similar questions on the other docs and other selector type

Reply:
> Must they? A lot of the docs seem leftover from the first draft where the fields were not mutually exclusive
> 
> I'd encourage general webhook authors to:

how would you like to phrase: "clients can send you data that is invalid that a kube-apiserver will never send you.  Think about it advance and don't crash"?

Reply:
I think this covers it?

> I'd encourage general webhook authors to:
> 
> * ensure rawSelector and requirements are not both set
> * consider the requirements field if set
> * not try to parse or consider the rawSelector field if set

Reply:
ok, willing to update.

Trying to decide how much carries to actual API documentation that appears on the SAR endpoint.  I haven't checked docs to see if we have a separate place yet.